### DISCUSSION TITLE
Using Custom Load Profile via API

### DISCUSSION THREAD

**Discussion by gkaminski** [NREL/REopt-API-Analysis#57] [8 - API: Application Programming Interface] [answered]
Hello,
I'm attempting to pass a custom load profile using the API and am getting a "Request Rejected" error returned. I am passing a list of 8760 float values. I have no problems when I comment out the "loads_kw" and "year" lines and uncomment the next three lines. Any guidance and/or example code would be greatly appreciated!
Thanks!

```
load_profile = [10000.0] * 8760

REoptURL = "https://developer.nrel.gov/api/reopt/stable"
REoptParams = {
    "Scenario" : {
        "Site" : {
            "latitude" : 40.73,
            "longitude" : 74.02,
            "roof_squarefeet" : 5000.0,
            "PV" : {
                "tilt" : 20.0, #Default for PVWatts
                "location" : "roof",
                "module_type" : 0, #standard, default
                "array_type" : 1, #Roof, fixed
                "inv_eff" : 0.96,
                "azimuth" : 180.0,
                "installed_cost_us_dollars_per_kw" : 1600.0 #default
                },
            "LoadProfile" : {
                "loads_kw" : load_profile,
                "critical_loads_kw" : load_profile,
                "year" : 2017
                #"doe_reference_name" : "RetailStore",
                #"annual_kwh" : 10000000.0,
                #"city" : "NewYork"
                },
            "Storage" : {
                "can_grid_charge" : False,
                "min_kw" : 130.0,
                "min_kwh" : 232.0,
                "soc_init_pct" : 0.0
                },
            "ElectricTariff" : {
                "urdb_label" : "5cd202ab5457a3b21b54e9d6" #https://apps.openei.org/USURDB/rate/view/5cd202ab5457a3b21b54e9d6
                }
            }
        }
    }
```

**Comment by NLaws** [comment]
That is odd. Try removing the comment after `urdb_label`; URLs in POSTs will lead to rejections.

**Reply by gkaminski** [reply]
Doesn't look like that helped unfortunately. Still get this error:
```
main         ERROR    Status code 404. b"<html><head><title>Request Rejected</title></head><body>An error has occurred. Please consult with your administrator.<br><br>Your support ID is: 18387969632454302270<br><br><a href='javascript:history.back();'>[Go Back]</a></body></html>"
```
It does seem like a formatting error though. I also just tried converting the whole list of values to a string and passing it that way, which at least gave a more descript error ("input_errors": ["Could not convert loads_kw ([10000.0, 10000.0,..., 10000.0]) in Scenario>Site>LoadProfile to \'float\'>"].

**Reply by NLaws** [reply]
Are you reusing the "inputs" from the API and do they contain a "urdb_response" ?  According to your support ID the request was rejected because of a series of 1's, I think in the "weekdayschedule" of the "urdb_response", which were identified as a potential SQL injection. I am working with our IT team to see if we can make a security exception for series of 1's but for now you can use the URDB tariff by just providing the "urdb_label" (like in your original post above).

**Reply by gkaminski** [reply]
That seemed to work. Thanks for the help!

And yes, I was also sending a "urdb_response" parameter. I had omitted this from my post since I only included the parameters initially defined in the scenario (and it worked when I wasn't passing a custom load profile).
I was sending the urdb_response similarly to the example in the documentation [here](https://github.com/NREL/REopt-API-Analysis/blob/master/single_scenario_example.ipynb).

```
load_electric_rate = "ConEd_SC-9"
with open(os.path.join(rates_path, load_electric_rate + ".json"), "r") as fp:
    rate_1 = json.load(fp)
REoptParams["Scenario"]["Site"]["ElectricTariff"]["urdb_response"] = rate_1
```
For reference, this is what I was sending as urdb_response:

`{"energyweekdayschedule":[[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]],"energyweekendschedule":[[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]],"energyratestructure":[[{"unit":"kWh","rate":0.0079,"adj":0.003985}]],"demandratestructure":[[{"rate":0}],[{"rate":12.43,"adj":11.09}],[{"rate":19.2}],[{"rate":28.15,"adj":11.09}]],"demandweekdayschedule":[[0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0],[0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0],[0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0],[0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0],[0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0],[0,0,0,0,0,0,0,0,3,3,3,3,3,3,3,3,3,3,2,2,2,2,0,0],[0,0,0,0,0,0,0,0,3,3,3,3,3,3,3,3,3,3,2,2,2,2,0,0],[0,0,0,0,0,0,0,0,3,3,3,3,3,3,3,3,3,3,2,2,2,2,0,0],[0,0,0,0,0,0,0,0,3,3,3,3,3,3,3,3,3,3,2,2,2,2,0,0],[0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0],[0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0],[0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0]],"demandweekendschedule":[[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]],"flatdemandstructure":[[{"rate":5.26,"adj":-0.66}],[{"rate":18.36,"adj":-0.66}]],"flatdemandmonths":[0,0,0,0,0,1,1,1,1,0,0,0],"fixedmonthlycharge":6.63}`

**Reply by NLaws** [reply]
yeah sorry about that; the list of things that can be rejected for security reasons is ever evolving.

**Comment by NLaws** [comment, marked as answer]
Answer is: the series of 1's in the `urdb_response.demandweekdayschedule` was flagged as a possible SQL injection. 
Solution: use `urdb_label` when possible (and we will see about making an exception for series of 1's in the `urdb_response`.
